python/revng/scripting/projects/project.py

```diff
@@ -0,0 +1,240 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import os
+import shutil
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from tempfile import TemporaryDirectory
+from typing import Dict, List
+
+from revng.model import Binary, StructBase, get_element_by_path  # type: ignore[attr-defined]
+from revng.tupletree import TypedList
+
+
+class Project(ABC):
+    """
+    This class is the basis for building a project. It defines
+    abstract methods that the inherited classes have to implement
+    and it also implements some common methods.
+    """
+
+    def __init__(self):
+        self.model: Binary = Binary()
+        self.last_saved_model: Binary = Binary()
+        self.pipeline_description: Dict = {}
+        self.input_binary: str
+        self.revng_executable: str
+        self.resume_dir: str
+        self._load()
+
+    @abstractmethod
+    def _load(self):
+        """
+        Load the model and pipeline description, this is used when you want to reload
+        from the resumed dir.
+        """
+        pass
+
+    @abstractmethod
+    def import_binary(self, input_binary: str):
+        """
+        Use this method to import the binary into the project. This method is a prerequisite
+        before running any analyses or getting artifacts.
+        """
+        pass
+
+    def get_artifact(self, artifact_name: str) -> bytes | str:
+        """
+        High level method that you can call to fetch an artifact without specifying targets.
+        """
+        return self._get_artifact(artifact_name)
```

Comment:
This is exactly the same as `_get_artifact`. Merge.

Reply:
This function does not have the `targets_class`, it's meant to be that way so that it's less confusing for end users

Reply:
That argument is optional. I find it more confusing.

Reply:
Reworked a bit how things are handled, right now there's `get_artifact` (the user facing one) and then `_get_artifact` (an `abstractmethod` that implements the actual logic on the implementation)